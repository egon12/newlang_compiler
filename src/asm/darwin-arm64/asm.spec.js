const fs = require('fs');
const asm = require('./asm');
const { Generator }  = require('./literal_asm');

it('should compile to asm', () => {
	const result = asm(node)
	expect(result).toMatchSnapshot();
	fs.writeFileSync(__dirname + '/asm.spec.s', result);
})

it('Generator should compile to asm', () => {
	const generator = new Generator();
	const result = generator.generate(node)
	expect(result).toMatchSnapshot();
})

let node = [
	{
		type: 'FunctionDefinition',
		namespace: 'main',
		name: 'main',
		parameters: [],
		body: [
			{
				type: 'VariableDefinition',
				name: 'a',
				value: {
					type: 'CallExpression',
					namespace: 'main',
					name: 'plusOne',
					parameters: [
						{ type: 'Number', value: "4" },
					],
				}
			},
			{
				type: 'CallStatement',
				value: {
					type: 'CallExpression',
					namespace: '',
					name: 'printint', // using stdio
					parameters: [
						{ type: 'Variable', id: 'a' },
					],
				}
			},
			{
				type: 'CallStatement',
				value: {
					type: 'CallExpression',
					namespace: '',
					name: 'println', // using stdio
					parameters: [],
				}
			},
			{
				type: 'ReturnStatement',
				value: {
					type: 'Number',
					value: 0,
				}
			}
		]
	},
	{
		type: 'FunctionDefinition',
		namespace: 'main',
		name: 'plusOne',
		parameters: [
			{ name: 'x', type: { type: 'BuiltInType', name: 'int' } },
		],
		body: [
			{
				type: 'ReturnStatement',
				value: {
					type: 'AddOperator',
					left: { type: 'Number', value: "1" },
					right: { type: 'Variable', id: 'x' },
				}
			}
		]
	},
	{
		type: 'FunctionDefinition',
		namespace: 'main',
		name: 'add',
		parameters: [
			{ name: 'a', type: { type: 'BuiltInType', name: 'int' } },
			{ name: 'b', type: { type: 'BuiltInType', name: 'int' } },
		],
		body: [
			{
				type: 'ReturnStatement',
				value: {
					type: 'AddOperator',
					left: { type: 'Variable', id: 'a' },
					right: { type: 'Variable', id: 'b' },
				}
			}
		]
	}
]
