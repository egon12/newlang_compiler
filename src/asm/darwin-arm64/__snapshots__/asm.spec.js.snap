// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should compile to asm 1`] = `
".global _main
.align 2
.extern _printf

_main:
	bl main.main
	mov x16, #1
	svc 0x80
main.main:
	sub sp, sp, #16
	stp x29, x30, [sp, #16]
	add x29, x29, #16
	mov x0, #4
	bl main.plusOne
	mov x0, x0
	str x0, [x29, #-8]
	ldr x0, [x29, #-8]
	bl printint
	mov x0, x0
	bl println
	mov x0, x0
	mov x0, #0
	ldp x29, x30, [sp, #16]
	add sp, sp, #16
	ret
main.plusOne:
	sub sp, sp, #16
	stp x29, x30, [sp, #16]
	add x29, x29, #16
	str x0, [x29, #-8]
	mov x0, #1
	ldr x1, [x29, #-8]
	add x0, x0, x1
	ldp x29, x30, [sp, #16]
	add sp, sp, #16
	ret
main.add:
	sub sp, sp, #16
	stp x29, x30, [sp, #16]
	add x29, x29, #16
	str x0, [x29, #-8]
	str x1, [x29, #-16]
	ldr x0, [x29, #-8]
	ldr x1, [x29, #-16]
	add x0, x0, x1
	ldp x29, x30, [sp, #16]
	add sp, sp, #16
	ret

printint:
	sub sp, sp, #16
	stp x29, x30, [sp, #16]
	add x29, sp, #16
	str x0, [sp, #0]
	adrp x0, intStr@page
	add x0, x0, intStr@pageoff
	bl _printf
	ldp x29, x30, [sp, #16]
	add sp, sp, #16
	ret

println:
	sub sp, sp, #16
	stp x29, x30, [sp, #16]
	add x29, sp, #16
	adrp x0, lnstr@page
	add x0, x0, lnstr@pageoff
	bl _printf
	ldp x29, x30, [sp, #16]
	add sp, sp, #16
	ret


.data
lnstr: .asciz "\\n"
intStr: .asciz "%d"
	"
`;
